code/datums/components/scope.dm

```diff
@@ -0,0 +1,151 @@
+/datum/component/scope
+	/// How far we can extend, with modifier of 1, up to our vision edge, higher numbers multiply.
+	var/range_modifier = 1
+	/// Fullscreen object we use for tracking the shots.
+	var/atom/movable/screen/fullscreen/sniper/tracker
+
+/datum/component/scope/Initialize(range_modifier)
+	if(!isgun(parent))
+		return COMPONENT_INCOMPATIBLE
+	src.range_modifier = range_modifier
+
+/datum/component/scope/Destroy(force, silent)
+	if(tracker)
+		stop_zooming(tracker.marksman)
+	return ..()
+
+/datum/component/scope/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/on_move)
+	RegisterSignal(parent, COMSIG_ITEM_AFTERATTACK_SECONDARY, .proc/on_secondary_afterattack)
+	RegisterSignal(parent, COMSIG_GUN_TRY_FIRE, .proc/on_gun_fire)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/on_examine)
+
+/datum/component/scope/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_MOVABLE_MOVED,
+		COMSIG_ITEM_AFTERATTACK_SECONDARY,
+		COMSIG_GUN_TRY_FIRE,
+		COMSIG_PARENT_EXAMINE,
+	))
+
+/datum/component/scope/process(delta_time)
+	if(!tracker.marksman.client)
+		stop_zooming(tracker.marksman)
+		return
+	if(!length(tracker.marksman.client.keys_held & tracker.marksman.client.movement_keys))
+		tracker.marksman.face_atom(tracker.given_turf)
+	animate(tracker.marksman.client, 0.2 SECONDS, easing = SINE_EASING, flags = EASE_OUT, pixel_x = tracker.given_x, pixel_y = tracker.given_y)
+
+/datum/component/scope/proc/on_move(atom/movable/source, atom/oldloc, dir, forced)
+	SIGNAL_HANDLER
+
+	if(!tracker)
+		return
+	stop_zooming(tracker.marksman)
+
+/datum/component/scope/proc/on_secondary_afterattack(datum/source, atom/target, mob/user, proximity_flag, click_parameters)
+	SIGNAL_HANDLER
+
+	if(tracker)
+		stop_zooming(user)
+	else
+		start_zooming(user)
+	return COMPONENT_SECONDARY_CANCEL_ATTACK_CHAIN
```

Comment:
Doesn't the beam rifle have a version of this that can be held? I think you were talking about wanting that.

Reply:
beam rifle handles everything on left click. handling stuff by holding right click and clicking left click is impossible, as mouseup never gets called again after you left click, so you never leave aim mode